File: backend/spring-boot-language-app/src/main/java/com/jackschaeffer/languageapp/entity/Song.java
package com.jackschaeffer.languageapp.entity;

import lombok.Data;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import javax.persistence.*;
import java.util.Date;
import java.util.Set;

@Entity
@Table(name="song")
@Data
public class Song {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "id")
    private Long id;

    @ManyToOne
    @JoinColumn(name = "artist_id", nullable = false)
    private Artist artist;

    @ManyToOne
    @JoinColumn(name = "genre_id", nullable = false)
    private Genre genre;

    @ManyToMany
    @JoinTable(
        name = "song_phrase",
        joinColumns = @JoinColumn(name = "song_id"),
        inverseJoinColumns = @JoinColumn(name = "phrase_id")
    )
    private Set<Phrase> phrases;

    @Column(name = "fr_title")
    private String frTitle;

    @Column(name = "en_title")
    private String enTitle;

    @Column(name = "description")
    private String description;

    @Column(name = "fr_lyrics")
    private String frLyrics;

    @Column(name = "en_lyrics")
    private String enLyrics;

    @Column(name = "image_url")
    private String imageUrl;

    @Column(name = "youtube_url")
    private String youtubeUrl;

    @Column(name = "release_year")
    private int releaseYear;

    @Column(name = "date_created")
    @CreationTimestamp
    private Date dateCreated;

    @Column(name = "last_updated")
    @UpdateTimestamp
    private Date lastUpdated;
}

File: frontend/angular-languageapp/src/app/common/song.ts
import { Artist } from "./artist";
import { Genre } from "./genre";
import { Phrase } from "./phrase";

export class Song {
    id!: number;
    frTitle!: string;
    enTitle!: string;
    description!: string;
    frLyrics!: string;
    enLyrics!: string;
    imageUrl!: string;
    youtubeUrl!: string;
    yearReleased!: number;
    dateCreated!: Date;
    lastUpdated!: Date;

    genre: Genre = new Genre();
    artist: Artist = new Artist();
    phrases: Phrase[] = [];

    frLyricsArr!: string[];
    enLyricsArr!: string[];

}

